feat: Support exception tagging for HTTP response

kavirajk opened the pull request:
<!---AI changelog entry and formatting assistance: false-->
### Changelog category (leave one):

- Backward Incompatible Change

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Support exception tagging for HTTP results response to enable clients to parse exceptions more reliably. Resolves #75175.  The setting `http_write_exception_in_output_format` is disabled by default for consistency across formats. 

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### Details

This PR adds a consistent and reliable way to return exceptions that occur mid-stream during HTTP request serving. The behavior is enabled by default and adds an X-ClickHouse-Exception-Tag header to every response.

When an exception occurs, it will be written in the following form which makes it easier for clients to parse the exception message reliably:

```
HTTP/1.1 200 OK
X-ClickHouse-Exception-Tag: PU1FNUFH98

bodybodybodybody
__exception__PU1FNUFH98
Big bam accrued right while reading the data
45 PU1FNUFH98__exception__
```

Here 45 is the size of the exception message, and PU1FNUFH98 is the exception token.

Resolves #75175

kavirajk committed:
feat: Support exception tagging for HTTP response

Fixes: https://github.com/ClickHouse/ClickHouse/issues/75175

This PR adds consistent and reliable way to return exceptions occured in
mid-stream during serving HTTP request.

It implements the proposal agreed on the attached issue. Everytime the
exception occurs, it will be written in following form (which makes it
easier for clients to parse the exception message reliably)

```
HTTP/1.1 200 OK↵
X-ClickHouse-Exception-Tag: PU1FNUFH98↵
↵
bodybodybodybody↵
__exception__PU1FNUFH98↵
Big bam accrued right while reading the data↵
45 PU1FNUFH98__exception__↵
```

Here 45 is the size of the Big bam accrued right while reading the data↵ message, and PU1FNUFH98 is the exception token.

Also this behavior change comes into play only when the settings
`http_exception_tagging` is 1 (default 0)
Otherwise it's a no-op

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

slvrtrn commented on src/Core/Settings.cpp:
```diff
@@ -2067,6 +2067,9 @@ Write exception in output format to produce valid output. Works with JSON and XM
     DECLARE(UInt64, http_response_buffer_size, 0, R"(
 The number of bytes to buffer in the server memory before sending a HTTP response to the client or flushing to disk (when http_wait_end_of_query is enabled).
+)", 0) \
+    DECLARE(Bool, http_exception_tagging, false, R"(
```
Let's also test if we can enable it for a readonly user

slvrtrn replied:
I am also curious: does it throw if it is set on an older CH version, e.g. do we need to track CH version on the language client side before setting it?

kavirajk replied:
Had a chat with @slabko and @tavplubix. We decided not to have this settings at all.

Given, we made sure we have a `\r\n` before between exception marker and a tag. It wouldn't break any existing behavior and this feature can be enabled by default without needing any toggle.

I will remove the this setting.

slvrtrn reviewed (commented):
@tavplubix PTAL

kavirajk committed:
chore: fix breaking style checks and stateless tests

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
chore: Remove the need for `http_exception_tagging` settings.

After a discussion with people, we agreed this can be default behavior
without needing any toggle by new settings.

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
chore: Fix failing tests by adding new Exception-Tag header

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
fix: Handle `X-ClickHouse-Exception-Tag` in couple of old tests

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
fix: Add missing newline on http_head.reference test

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
restore `tests/queries/0_stateless/00501_http_head.reference`

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

tavplubix commented on tests/queries/0_stateless/03417_http_exception_tag.sh:
```diff
@@ -0,0 +1,55 @@
+
+echo "Test 2: Verify exception tag is alphanumeric"
+TAG=$(${CLICKHOUSE_CURL} -sS -D - "${CLICKHOUSE_URL}" --data-binary "SELECT 1" | grep "X-ClickHouse-Exception-Tag" | cut -d':' -f2 | tr -d ' \r\n')
+if [[ "$TAG" =~ ^[A-Za-z0-9]+$ ]] && [[ ${#TAG} -eq 8 ]]; then
```
I've noticed that getRandomASCIIString (unlike FunctionRandomPrintableASCII) only uses `[a-z]`, which means 26^8 = 37.6 bits of entropy, which is probably good enough, but it's much less than UUIDv4 has, for example

tavplubix replied:
Why should we care about collisions between different requests?

The main purposeof having a random string is to be able to distinguish the exception marker from the data in the response. It's already quite unlikely to find the same random string in the data, but why not make it longer?

tavplubix commented on src/Server/HTTP/WriteBufferFromHTTPServerResponse.cpp:
```diff
@@ -104,6 +105,8 @@ void WriteBufferFromHTTPServerResponse::finishSendHeaders()
     if (add_cors_header)
         response.set("Access-Control-Allow-Origin", "*");
 
+    response.set("X-ClickHouse-Exception-Tag", exception_tag);
```
Should we add it to `Access-Control-Expose-Headers`?

tavplubix commented on tests/queries/0_stateless/03417_http_exception_tag.reference:
```diff
@@ -0,0 +1,11 @@
+Test 1: Check that X-ClickHouse-Exception-Tag header is present
+X-ClickHouse-Exception-Tag:
```
`clickhouse-test` doesn't like `Exception` substring in stdout

tavplubix committed:
fix tests

kavirajk committed:
PR remarks with following changes

1. Remove unused `getExceptionTag()`
2. Add exception tag to `Access-Control-Expose-Headers`
3. Calculate the proper trim size of exception message. Making sure the
   whole exception block doesn't exceed MAX_EXCEPTION_SIZE
4. Update the tests accordingly

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
fix http_head test

Needed as we added exception tag header X-ClickHouse-Exception-Tag in Access-Control-Expose-Headers

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
Fix fast-tests by adding new `X-ClickHouse-Exception-Tag` headers

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
Make the exception tag test deterministic.

1. Uses the http buffering, response size, block size consistently
2. Remove any additional LIMIT clause

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

tavplubix committed:
try to make the test more stable

tavplubix commented on src/Server/HTTP/WriteBufferFromHTTPServerResponse.cpp:
```diff
@@ -322,12 +328,41 @@ bool WriteBufferFromHTTPServerResponse::cancelWithException(HTTPServerRequest &
+
+            // 2 bytes represents - \r\n
+            // 1 byte represents - ' ' (space between <message_length> <TAG>) in the above exception block format
+            size_t size_message_excluded = 2 + EXCEPTION_MARKER.size() + 2 + EXCEPTION_TAG_LENGTH + 2 + sizeof(size_t) + 1 + EXCEPTION_TAG_LENGTH + 2 + EXCEPTION_MARKER.size() + 2;
```
> but limited_message.size() return size_t and the actual size depends on the arch no? say 4-byte in 32-bit machine and 8 byte in 64-bit machine.

Yes, you're right, but we don't build clickhouse for 32-bit architectures, it's not supported

However, if we did - this code would be incorrect:
 - sizeof(size_t) would be 4, but this is for binary representation of the number
 - limited_message.size() can be, for example, 10000 (it's still way below the limit), but you need 5 bytes to write this number as string because there are 5 decimal digits, and it doesn't fit into 4 bytes 

So this would be a bug

kavirajk replied:
Make sense. Updated to 8 bytes.

kavirajk committed:
Couple of PR remarks

1. Change exception tag size to 16 bytes
2. Hardcode the exception size to 8 bytes

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

tavplubix committed:
Change http_write_exception_in_output_format to false

tavplubix committed:
Merge branch 'master' into kavirajk/support-exception-tag-consistent-error-handling

tavplubix committed:
Update settings changes history with new values

Updated settings changes history to reflect new and modified settings.

tavplubix committed:
Merge branch 'master' into kavirajk/support-exception-tag-consistent-error-handling

tavplubix commented:
Could you please update docs as well?

kavirajk committed:
exception tag test accepting changes from upstream

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
Handle zero-chunk on exception and fix JSON test case

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk commented:
@tavplubix when implementing the support on client side (in Golang client), I found a subtle bug. 

**Bug:** Currently during exception, CH server doesn't send `zero-chunk` at the end. Which breaks the [HTTP RFC 7230 Chunked Transfer Encoding protocol.](https://datatracker.ietf.org/doc/html/rfc7230#section-4.1)

Steps to reproduce with CH server built on this PR

### Notice there is no zero chunk `0\r\n` at the end
```
$ curl -i --raw -s --max-time 120 \
        'http://localhost:8123/?database=default&wait_end_of_query=0&max_threads=1' \
        --data-binary "SELECT throwIf(number=3, 'there is a exception') FROM system.numbers SETTINGS max_block_size=1 FORMAT Native" \
        --output broken_response.raw

$ tail -c 60 broken_response.raw | hexdump -c
0000000       (   v   e   r   s   i   o   n       2   5   .   1   1   .
0000010   1   .   1   )  \n   3   0   3       r   w   z   h   v   n   a
0000020   m   h   q   c   f   y   f   b   h  \r  \n   _   _   e   x   c
0000030   e   p   t   i   o   n   _   _  \r  \n  \r  \n
000003c

$ tail -c 60 broken_response.raw | cat -v
 (version 25.11.1.1)
303 rwzhvnamhqcfyfbh^M
__exception__^M
^M
```

Go client (like any standard HTTP client), fail to parse the last chunk(returning EOF early), but it won't return the "partial chunk" because it didn't find the last `zero-chunk`. Hence we cannot extract the "exception message" from the partial chunk. You can find the [demo_zero_chunk.go here](https://gist.github.com/kavirajk/102b90118dfca1e4eb52cc9ae9a79c04)

```
$ go run demo_zero_chunk.go
❌ Error reading body: unexpected EOF
   This happens because ClickHouse doesn't send proper chunk terminator!
   Captured 0 bytes before error
exit status 1
```

### I fixed the flush code and you see the zero-chunk in the `hexdump` and `cat -v`
```
$ curl -i --raw -s --max-time 120 \
        'http://localhost:8123/?database=default&wait_end_of_query=0&max_threads=1' \
        --data-binary "SELECT throwIf(number=3, 'there is a exception') FROM system.numbers SETTINGS max_block_size=1 FORMAT Native" \
        --output correct_response.raw

$ tail -c 60 correct_response.raw | hexdump -c
0000000   s   i   o   n       2   5   .   1   1   .   1   .   1   )  \n
0000010   3   0   3       a   i   o   z   l   b   c   t   b   f   l   s
0000020   f   g   d   o  \r  \n   _   _   e   x   c   e   p   t   i   o
0000030   n   _   _  \r  \n  \r  \n   0  \r  \n  \r  \n
000003c

$ tail -c 60 correct_response.raw | cat -v
sion 25.11.1.1)
303 aiozlbctbflsfgdo^M
__exception__^M
^M
0^M
^M

```

Also Go client is happy as well :tada: 

```
$ go run demo_zero_chunk.go
✅ Successfully read complete response (555 bytes)

⚠️  Query completed with exception (streamed in body)
Exception: Code: 395. DB::Exception: there is a exception: while executing 'FUNCTION throwIf(equals(__table1.number, 3_UInt8) :: 0, 'there is a exception'_String :: 2) -> throwIf(equals(__table1.number, 3_UInt8), 'there is a exception'_String) UInt8 : 1'. (FUNCTION_THROW_IF_VALUE_IS_NON_ZERO) (version 25.11.1.1)
exit status 1
```


[Here is the full (draft) Go client support implementation.](https://github.com/ClickHouse/clickhouse-go/pull/1694)

Let me know what you think?

tavplubix commented:
> **Bug**: Currently during exception, CH server doesn't send `zero-chunk` at the end.

ClickHouse does this intentionally. See https://github.com/ClickHouse/ClickHouse/pull/68800

kavirajk committed:
revert handling zero-chunk in the previous commit

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
chore: reverting exception tag test file to working one.

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
fix JSON tests cases enable `http_write_exception_in_output_format`

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
update the HTTP interface docs

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
fix typos in http docs as per aspell style checks

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
fix docs linter errors

1. Hard tabs -> spaces
2. Context for shell commands

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
doc: Clarify `X-ClickHouse-Exception-Tag`

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>
